File: android/app/src/main/java/com/lilru/liftr/ui/pets/ProfilePetCombatSection.kt
package com.lilru.liftr.ui.pets

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Info
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Text
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import com.lilru.liftr.R
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import coil.compose.SubcomposeAsyncImage
import com.lilru.liftr.data.PetCombatPetSummaryWire
import com.lilru.liftr.data.PetCombatStatsSummaryWire
import com.lilru.liftr.data.PetService

@Composable
fun PetCombatSummaryImage(
    pet: PetCombatPetSummaryWire,
    modifier: Modifier = Modifier,
    height: Dp = 120.dp,
    contentPadding: Dp = 0.dp
) {
    val imageUrl = PetService.resolvedCombatPetImageUrl(pet)

    SubcomposeAsyncImage(
        model = imageUrl,
        contentDescription = null,
        modifier = modifier
            .height(height)
            .padding(contentPadding),
        contentScale = ContentScale.Fit,
        loading = {
            CircularProgressIndicator(modifier = Modifier.height(height))
        }
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun PetCombatComparisonStats(
    attackerName: String,
    defenderName: String,
    attackerLevel: Int,
    defenderLevel: Int,
    attackerStats: PetCombatStatsSummaryWire,
    defenderStats: PetCombatStatsSummaryWire,
    modifier: Modifier = Modifier
) {
    var showStatCombatHelp by rememberSaveable { mutableStateOf(false) }
    val statCombatHelpSheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(6.dp)) {
        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
            IconButton(onClick = { showStatCombatHelp = true }) {
                Icon(
                    Icons.Filled.Info,
                    contentDescription = stringResource(R.string.pet_stat_combat_info_content_description)
                )
            }
        }
        Row(Modifier.fillMaxWidth()) {
            Text(attackerName, modifier = Modifier.weight(1f), style = MaterialTheme.typography.labelSmall, maxLines = 1)
            Text("vs", style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
            Text(defenderName, modifier = Modifier.weight(1f), style = MaterialTheme.typography.labelSmall, maxLines = 1)
        }
        statRow("Level", attackerLevel, defenderLevel)
        statRow("Health", attackerStats.health, defenderStats.health)
        statRow("Strength", attackerStats.strength, defenderStats.strength)
        statRow("Defense", attackerStats.defense, defenderStats.defense)
        statRow("Speed", attackerStats.speed, defenderStats.speed)
        statRow("Intelligence", attackerStats.intelligence, defenderStats.intelligence)
        statRow("Agility", attackerStats.agility, defenderStats.agility)
        statRow("Stamina", attackerStats.stamina, defenderStats.stamina)
        statRow("Crit", attackerStats.criticalRate, defenderStats.criticalRate)
        statRow("Resistance", attackerStats.resistance, defenderStats.resistance)
        statRow("Explore", attackerStats.exploration, defenderStats.exploration)
        statRow("Happiness", attackerStats.happiness, defenderStats.happiness)
    }

    if (showStatCombatHelp) {
        ModalBottomSheet(
            onDismissRequest = { showStatCombatHelp = false },
            sheetState = statCombatHelpSheetState
        ) {
            PetStatCombatHelpSheetContent(onClose = { showStatCombatHelp = false })
        }
    }
}

@Composable
private fun statRow(label: String, left: Int, right: Int) {
    Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        Text("$left", modifier = Modifier.weight(1f), style = MaterialTheme.typography.labelSmall)
        Text(label, modifier = Modifier.weight(1f), style = MaterialTheme.typography.labelSmall, color = MaterialTheme.colorScheme.onSurfaceVariant)
        Text("$right", modifier = Modifier.weight(1f), style = MaterialTheme.typography.labelSmall)
    }
}

File: Liftr/Pet/ProfilePetCombatSection.swift
import SwiftUI

struct PetCombatSummaryImage: View {
    let pet: PetCombatPetSummary
    var height: CGFloat = 120
    var padding: CGFloat = 0

    var body: some View {
        AsyncImage(url: pet.imageURL) { phase in
            switch phase {
            case .empty:
                ProgressView()
                    .frame(height: height)
            case .success(let image):
                image
                    .resizable()
                    .scaledToFit()
                    .padding(padding)
                    .frame(height: height)
            case .failure:
                Image(systemName: "pawprint.fill")
                    .font(.largeTitle)
                    .foregroundStyle(.secondary)
                    .frame(height: height)
            @unknown default:
                EmptyView()
            }
        }
    }
}

struct PetCombatComparisonStats: View {
    let attackerName: String
    let defenderName: String
    let attackerStats: PetCombatStatsSummary
    let defenderStats: PetCombatStatsSummary
    let attackerLevel: Int
    let defenderLevel: Int

    var body: some View {
        VStack(spacing: 8) {
            HStack {
                Text(attackerName)
                    .font(.caption2.weight(.semibold))
                    .lineLimit(1)
                Spacer()
                Text("vs")
                    .font(.caption2)
                    .foregroundStyle(.secondary)
                Spacer()
                Text(defenderName)
                    .font(.caption2.weight(.semibold))
                    .lineLimit(1)
            }

            statCompareRow(label: "Level", left: attackerLevel, right: defenderLevel)
            statCompareRow(label: "Health", left: attackerStats.health, right: defenderStats.health)
            statCompareRow(label: "Strength", left: attackerStats.strength, right: defenderStats.strength)
            statCompareRow(label: "Defense", left: attackerStats.defense, right: defenderStats.defense)
            statCompareRow(label: "Speed", left: attackerStats.speed, right: defenderStats.speed)
            statCompareRow(label: "Intelligence", left: attackerStats.intelligence, right: defenderStats.intelligence)
            statCompareRow(label: "Agility", left: attackerStats.agility, right: defenderStats.agility)
            statCompareRow(label: "Stamina", left: attackerStats.stamina, right: defenderStats.stamina)
            statCompareRow(label: "Crit", left: attackerStats.criticalRate, right: defenderStats.criticalRate)
            statCompareRow(label: "Resistance", left: attackerStats.resistance, right: defenderStats.resistance)
            statCompareRow(label: "Explore", left: attackerStats.exploration, right: defenderStats.exploration)
            statCompareRow(label: "Happiness", left: attackerStats.happiness, right: defenderStats.happiness)
        }
    }

    private func statCompareRow(label: String, left: Int, right: Int) -> some View {
        HStack(spacing: 8) {
            Text("\(left)")
                .font(.caption2.monospacedDigit())
                .frame(maxWidth: .infinity, alignment: .leading)
                .foregroundStyle(left >= right ? .primary : .secondary)
            Text(label)
                .font(.caption2)
                .foregroundStyle(.tertiary)
                .frame(width: 72)
            Text("\(right)")
                .font(.caption2.monospacedDigit())
                .frame(maxWidth: .infinity, alignment: .trailing)
                .foregroundStyle(right >= left ? .primary : .secondary)
        }
    }
}
